### PR TITLE
Fixed notice 'Undefined index: reduction_applies'

### LINKED ISSUE #25650: BO - Order page - Debug mode enabled - an exception is displayed when adding a cheapest product or removing it from the Order with a specific cart rule (Exclude discounted products enabled)
#### Describe the bug

We create a cart rule with a percent discount applied on the cheapest product and Exclude discounted products option is enabled
In the BO > Orders > Order details page, an exception is displayed when we try to add/remove a cheaper product than the ordered products.

Only when debug mode is enabled.

#### Steps to Reproduce

1. Create an Order with one product
2. Create a cart rule:
- no code, no condition
- Discount percent + cheapest product checked + exclude discounted products enabled  
image

3. Go to the last order created and try to add another product to the order cheaper than the product ordered
4. see error 
image
5.  Refresh page => ok 
image
5. remove the cheapest product => see the error
6. refresh the page => ok

1777 + debug mode enabled => NOK: https://drive.google.com/file/d/1hKxmyPisjc0Op3Bzx9M6g3Xsj8xRLDZI/view?usp=sharing
1777+ Debug mode off => ok: https://drive.google.com/file/d/1L5Ehc8Uo4et7JkfihBh23o7RkmlgyNRg/view?usp=sharing
1769 => OK: https://drive.google.com/file/d/1V14OUoEKpTb9fMMsCxXsvsAhu9yDyri9/view?usp=sharing

#### Additional information

* PrestaShop version: 1777, 1776, 1770
* PHP version: 7.2

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Fixed notice 'Undefined index: reduction_applies'
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #25650
| How to test?      | Cf. #25650

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27795)

### CODE REVIEW

File: src/Core/Cart/CartRuleCalculator.php [hunk 1]
```diff
@@ -153,7 +153,7 @@ protected function applyCartRule(CartRuleData $cartRuleData, $withFreeShipping =
                 $cartRowCheapest = null;
                 foreach ($this->cartRows as $cartRow) {
                     $product = $cartRow->getRowData();
-                    if (((($cartRule->reduction_exclude_special && !$product['reduction_applies'])
+                    if (((($cartRule->reduction_exclude_special && isset($product['reduction_applies']) && !$product['reduction_applies'])
```
Comment (saulaski) [inline src/Core/Cart/CartRuleCalculator.php]: Only a suggestion, not required
```suggestion
                    if (((($cartRule->reduction_exclude_special && !($product['reduction_applies'] ?? true)
```

**Review by saulaski — DISMISSED** [review]

File: src/Core/Cart/CartRuleCalculator.php (hunk 1 above)
Comment (Progi1984) [reply src/Core/Cart/CartRuleCalculator.php]: I like it ;)

**Review by matthieu-rolland — DISMISSED** [review]

**Review by saulaski — DISMISSED** [review]

**Review by jolelievre — DISMISSED** [review]
Thanks @Progi1984

**Review by PierreRambaud — DISMISSED** [review]

**Review by tegbessou — DISMISSED** [review]

**Review by marwachelly — CHANGES_REQUESTED** [review]
Hello, 

The display of the discount is missing. Could you please check?
Actual behavior:
image

Expected behavior:
image

Thanks,

**Comment by Progi1984** [comment]
@marwachelly I just come to check (with the latest develop) and I haven't the problem : 
image

Could you check again ?

**Comment by Progi1984** [comment]
@PierreRambaud @jolelievre @matthieu-rolland Rebased.

**Review by saulaski — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Review by marwachelly — DISMISSED** [review]
Hello,

I retested the PR but I still have the same issue as it is shon in the screen record bellow:

https://watch.screencastify.com/v/MOYx9F2NTgzGo33weNTy
image

Could you please check?

Thanks,

**Review by jolelievre — COMMENTED** [review]
@Progi1984 I don't know if it's reviewable yet but PHPStan is screaming ^^

File: tests/Integration/Behaviour/Features/Context/Domain/CartRuleFeatureContext.php [hunk 2]
```diff
@@ -103,7 +103,7 @@ public function createCartRuleWithName(array $properties, string $nameInDefaultL
 
     /**
      * @Given /^I specify (cart rule|its) "([^"]+)" as "([^"]+)"$/
-     * @Given /^I specify that (its) "([^"]+)" is "([^"]+)"$/
+     * @Given /^I specify that (its) "([^"]+)" is "([^"]*)"$/
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CartRuleFeatureContext.php:106]: I don't understand this one though, the second value cannot be empty so it should be a `+` at list one character must be found

File: tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature [hunk 3]
```diff
@@ -770,3 +769,55 @@ Feature: Order from Back Office (BO)
       | total_paid_real          | 7.420  |
       | total_shipping_tax_excl  | 7.000  |
       | total_shipping_tax_incl  | 7.420  |
+
+  @order-cart-rules1
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature]:
```suggestion
```

Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature]: This is only used to run the specific scenario during development/debug it should be cleaned

File: tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature [hunk 4]
```diff
@@ -770,3 +769,55 @@ Feature: Order from Back Office (BO)
...
+    And I specify that its active until "2040-01-01 12:00:00"
+    And I specify that its "quantity" is "100"
+    And I specify that its "quantity per user" is "100"
+    And I specify that its "priority" is "1"
+    And its minimum purchase amount in currency "USD" is "0.0"
+    And its minimum purchase amount is tax included
+    And its minimum purchase amount is shipping included
+    And I specify its status as enabled
+    And it gives a percentage reduction of "50" which excludes discounted products and applies to cheapest product
+    When I save it
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature]:
```suggestion
    Then I save it
```

Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature]: Just a suggestion but it reads better this way IMHO

File: tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature [hunk 5]
```diff
@@ -770,3 +769,55 @@ Feature: Order from Back Office (BO)
...
+    ## Add the product to the order
+    When I add products to order "bo_order1" without invoice and the following products details:
+      | name          | product_cheap     |
+      | amount        | 1                 |
+      | price         | 10                |
+    Then order "bo_order1" should have 2 products in total
+    And order "bo_order1" should have 1 cart rule
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature:822]: Maybe you should add a bit more details to check the cart rule values? Like cart rule amount and order details with discount values

**Review by jolelievre — COMMENTED** [review]
Thanks @Progi1984 

The fix is good for me, just suggestions of improvements for behat tests And I don't think the regexp modification is useful 🤔

File: tests/Integration/Behaviour/Features/Context/Domain/CartRuleFeatureContext.php (hunk 2 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Context/Domain/CartRuleFeatureContext.php:106]: That allows to define empty value (`+` = >= 1 character / `*` = >= 0 character).

File: tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature (hunk 3 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature]: Oopsie

File: tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature (hunk 4 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature]: Fixed

File: tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature (hunk 5 above)
Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature:822]: Fixed

File: tests/Integration/Behaviour/Features/Context/Domain/CartRuleFeatureContext.php (hunk 2 above)
Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Context/Domain/CartRuleFeatureContext.php:106]: Oh right I assumed there was always a value to set but I didn't think about setting an empty value on purpose ^^

File: tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature (hunk 3 above)
Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature]: I believe you forgot to remove this tag ^^ The it's all good to me

Comment (Progi1984) [reply tests/Integration/Behaviour/Features/Scenario/Order/order_cart_rules.feature]: @jolelievre Fixed. Oopsie².

**Review by jolelievre — APPROVED** [review]
Thanks @Progi1984

**Review by kpodemski — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Review by marwachelly — APPROVED** [review]
Hello,

PR is tested and the issue is well fixed.

https://user-images.githubusercontent.com/89604350/161785334-8a6246e9-cbe9-4d17-92ce-fcf578c35de3.mp4

Thanks,

**Comment by Progi1984** [comment]
Thanks @marwachelly

**Pull request merged**